java/pypowsybl/src/main/java/com/powsybl/python/network/NetworkCFunctions.java

```diff
@@ -1189,16 +1197,88 @@ private static void applyFixedPositions(DataframePointer fixedPositions, NadPara
         }
     }
 
+    record CustomBranchLabels(String side1, String side2, String middle, EdgeInfo.Direction arrow1, EdgeInfo.Direction arrow2) {
+    }
+
+    private static String getValueFromSeriesOrNull(StringSeries series, int row) {
+        return (series != null) ? series.get(row) : null;
+    }
+
+    private static EdgeInfo.Direction getDirectionFromSeriesOrNull(StringSeries series, int row) {
+        if (series == null) {
+            return null;
+        }
+        String dir = series.get(row);
+        return (dir != null && !dir.isEmpty()) ? EdgeInfo.Direction.valueOf(dir) : null;
+    }
+
+    private static Map<String, CustomBranchLabels> getNadCustomBranchLabels(int rowCount, StringSeries idSeries,
+                                                                            StringSeries side1Label, StringSeries side2Label,
+                                                                            StringSeries middleLabel, StringSeries arrow1,
+                                                                            StringSeries arrow2) {
+        Map<String, CustomBranchLabels> nadCustomBranchLabels = new HashMap<>();
+        for (int i = 0; i < rowCount; i++) {
+            String id = idSeries.get(i);
+            CustomBranchLabels labels = new CustomBranchLabels(
+                    getValueFromSeriesOrNull(side1Label, i),
+                    getValueFromSeriesOrNull(side2Label, i),
+                    getValueFromSeriesOrNull(middleLabel, i),
+                    getDirectionFromSeriesOrNull(arrow1, i),
+                    getDirectionFromSeriesOrNull(arrow2, i)
+            );
+            nadCustomBranchLabels.put(id, labels);
+        }
+        return nadCustomBranchLabels;
+    }
+
+    private static LabelProviderFactory getCustomLabelProviderFactory(Map<String, CustomBranchLabels> branchLabels) {
+        return (network, svgParameters) -> new DefaultLabelProvider(network, svgParameters) {
+            @Override
+            public Optional<EdgeInfo> getEdgeInfo(Graph graph, BranchEdge edge, BranchEdge.Side side) {
+                CustomBranchLabels bl = branchLabels.get(edge.getEquipmentId());
+                String label = null;
+                EdgeInfo.Direction arrowDirection = null;
+                if (bl != null) {
+                    label = side == BranchEdge.Side.ONE ? bl.side1 : bl.side2;
+                    arrowDirection = side == BranchEdge.Side.ONE ? bl.arrow1 : bl.arrow2;
+                }
+                return Optional.of(new EdgeInfo("ActivePower", arrowDirection, null, label));
+            }
+
+            @Override
+            public String getLabel(Edge edge) {
+                CustomBranchLabels bl = branchLabels.get(edge.getEquipmentId());
+                return (bl != null) ? bl.middle : null;
+            }
+        };
```

Comment:
Could you open a PR on powsybl-diagram to add such a LabelProvider, with a constructor based on the given map? So that after next powsybl-diagram release we remove this code here.